## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#7ada9209-03da-4945-8fac-81bc3f96883b) |
| mint-large-bucket.sh | :x: [more...](#ea2a5b58-7727-4729-9430-47c960a19b3b) |
| mint-compression-dist-xl.sh | :x: [more...](#d477e086-7fd0-4796-8901-f5de96cf1a24) |
<a name="7ada9209-03da-4945-8fac-81bc3f96883b"></a>
### 8675-28ae792/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32582
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8fc9f06a5a9f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 12 seconds
{
  "name": "awscli",
  "duration": 10514,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="d477e086-7fd0-4796-8901-f5de96cf1a24"></a>
### 8675-28ae792/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31267
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp feb8910eeed8:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 15 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 3 seconds
(4/13) Running mc tests ... done in 30 seconds
(5/13) Running minio-go tests ... done in 1 minutes and 30 seconds
(6/13) Running minio-js tests ... done in 59 seconds
(7/13) Running s3cmd tests ... done in 30 seconds
(8/13) Running aws-sdk-go tests ... done in 1 seconds
(9/13) Running aws-sdk-php tests ... done in 45 seconds
(10/13) Running healthcheck tests ... done in 0 seconds
(11/13) Running minio-dotnet tests ... done in 29 seconds
(12/13) Running minio-py tests ... FAILED in 2 minutes and 59 seconds
{
  "status": "FAIL",
  "duration": 150648,
  "args": {
    "object_name": "dfbe57f0-d819-4e6d-9ccd-3b39478d60ac_0 ~ dfbe57f0-d819-4e6d-9ccd-3b39478d60ac_1000",
    "bucket_name": "minio-py-test-e68d7d4a-1d5f-446b-9b9e-e2f8a31d6611"
  },
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1172, in test_list_objects_with_1001_files\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 417, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1949, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2033, in main\n    test_list_objects_with_1001_files(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1174, in test_list_objects_with_1001_files\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n",
  "function": "list_objects(bucket_name, prefix, recursive)",
  "name": "minio-py:test_list_objects_with_1001_files",
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty."
}

Executed 11 out of 13 tests successfully.
```
<a name="ea2a5b58-7727-4729-9430-47c960a19b3b"></a>
### 8675-28ae792/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30352
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ba4e0a4033e3:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 7 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 1 minutes and 11 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 24 seconds
{
  "args": {},
  "duration": 93548,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/admin-handlers-users.go
+++ b/cmd/admin-handlers-users.go
@@ -69,6 +69,16 @@ func (a adminAPIHandlers) RemoveUser(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	accessKey := vars["accessKey"]
 
+	ok, err := globalIAMSys.IsTempUser(accessKey)
+	if err != nil {
+		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+		return
+	}
+	if ok {
+		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, errIAMActionNotAllowed), r.URL)
+		return
+	}
+
 	if err := globalIAMSys.DeleteUser(accessKey); err != nil {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
@@ -516,6 +526,18 @@ func (a adminAPIHandlers) SetPolicyForUserOrGroup(w http.ResponseWriter, r *http
 	if globalWORMEnabled {
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrMethodNotAllowed), r.URL)
 		return
+	}
+
+	if !isGroup {
+		ok, err := globalIAMSys.IsTempUser(entityName)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		if ok {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, errIAMActionNotAllowed), r.URL)
+			return
+		}
 	}
 
 	if err := globalIAMSys.PolicyDBSet(entityName, policyName, isGroup); err != nil {

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -600,18 +600,38 @@ func (sys *IAMSys) ListUsers() (map[string]madmin.UserInfo, error) {
 	}
 
 	for k, v := range sys.iamUsersMap {
-		users[k] = madmin.UserInfo{
-			PolicyName: sys.iamUserPolicyMap[k].Policy,
-			Status: func() madmin.AccountStatus {
-				if v.IsValid() {
-					return madmin.AccountEnabled
-				}
-				return madmin.AccountDisabled
-			}(),
+		if !v.IsTemp() {
+			users[k] = madmin.UserInfo{
+				PolicyName: sys.iamUserPolicyMap[k].Policy,
+				Status: func() madmin.AccountStatus {
+					if v.IsValid() {
+						return madmin.AccountEnabled
+					}
+					return madmin.AccountDisabled
+				}(),
+			}
 		}
 	}
 
 	return users, nil
+}
+
+// IsTempUser - returns if given key is a temporary user.
+func (sys *IAMSys) IsTempUser(name string) (bool, error) {
+	objectAPI := newObjectLayerWithoutSafeModeFn()
+	if objectAPI == nil {
+		return false, errServerNotInitialized
+	}
+
+	sys.RLock()
+	defer sys.RUnlock()
+
+	creds, found := sys.iamUsersMap[name]
+	if !found {
+		return false, errNoSuchUser
+	}
+
+	return creds.IsTemp(), nil
 }
 
 // GetUserInfo - get info on a user.
@@ -634,6 +654,10 @@ func (sys *IAMSys) GetUserInfo(name string) (u madmin.UserInfo, err error) {
 	creds, found := sys.iamUsersMap[name]
 	if !found {
 		return u, errNoSuchUser
+	}
+
+	if creds.IsTemp() {
+		return u, errIAMActionNotAllowed
 	}
 
 	u = madmin.UserInfo{
@@ -670,6 +694,10 @@ func (sys *IAMSys) SetUserStatus(accessKey string, status madmin.AccountStatus) 
 	cred, ok := sys.iamUsersMap[accessKey]
 	if !ok {
 		return errNoSuchUser
+	}
+
+	if cred.IsTemp() {
+		return errIAMActionNotAllowed
 	}
 
 	uinfo := newUserIdentity(auth.Credentials{
@@ -719,9 +747,15 @@ func (sys *IAMSys) SetUser(accessKey string, uinfo madmin.UserInfo) error {
 		return errServerNotInitialized
 	}
 
+	cr, ok := sys.iamUsersMap[accessKey]
+	if cr.IsTemp() && ok {
+		return errIAMActionNotAllowed
+	}
+
 	if err := sys.store.saveUserIdentity(accessKey, false, u); err != nil {
 		return err
 	}
+
 	sys.iamUsersMap[accessKey] = u.Credentials
 
 	// Set policy if specified.
@@ -794,9 +828,12 @@ func (sys *IAMSys) AddUsersToGroup(group string, members []string) error {
 
 	// Validate that all members exist.
 	for _, member := range members {
-		_, ok := sys.iamUsersMap[member]
+		cr, ok := sys.iamUsersMap[member]
 		if !ok {
 			return errNoSuchUser
+		}
+		if cr.IsTemp() {
+			return errIAMActionNotAllowed
 		}
 	}
 
@@ -856,9 +893,12 @@ func (sys *IAMSys) RemoveUsersFromGroup(group string, members []string) error {
 
 	// Validate that all members exist.
 	for _, member := range members {
-		_, ok := sys.iamUsersMap[member]
+		cr, ok := sys.iamUsersMap[member]
 		if !ok {
 			return errNoSuchUser
+		}
+		if cr.IsTemp() {
+			return errIAMActionNotAllowed
 		}
 	}
 

--- a/pkg/auth/credentials.go
+++ b/pkg/auth/credentials.go
@@ -117,6 +117,11 @@ func (cred Credentials) IsExpired() bool {
 	return cred.Expiration.Before(time.Now().UTC())
 }
 
+// IsTemp - returns whether credential is temporary or not.
+func (cred Credentials) IsTemp() bool {
+	return cred.SessionToken != ""
+}
+
 // IsValid - returns whether credential is valid or not.
 func (cred Credentials) IsValid() bool {
 	// Verify credentials if its enabled or not set.

